modules/sde-external-services/edc/src/main/java/org/eclipse/tractusx/sde/edc/services/ConsumerControlPanelService.java

```diff
@@ -234,4 +248,114 @@ public void subscribeDataOffers(ConsumerRequest consumerRequest, String processI
 
 	}
 
+	@SneakyThrows
+	public EDRCachedResponse verifyOrCreateContractNegotiation(String connectorId,
+			Map<String, String> extensibleProperty, String recipientURL, ActionRequest action, OfferRequest offer) {
+		// Verify if there already EDR process initiated then skip t for again download
+		List<EDRCachedResponse> eDRCachedResponseList = edrRequestHelper.getEDRCachedByAsset(offer.getAssetId());
+		EDRCachedResponse checkContractNegotiationStatus = verifyEDRResponse(eDRCachedResponseList);
+
+		if (checkContractNegotiationStatus == null) {
+			log.info("There was no EDR process initiated or may be EDR token was expired " + offer.getAssetId()
+					+ ", so initiating EDR process");
+			edrRequestHelper.edrRequestInitiate(recipientURL, connectorId, offer.getOfferId(), offer.getAssetId(),
+					action, extensibleProperty);
+		} else {
+			log.info("There was EDR process initiated " + offer.getAssetId() + ", so ignoring EDR process initiation");
+		}
+
+		if (checkContractNegotiationStatus == null || !NEGOTIATED.equals(checkContractNegotiationStatus.getEdrState()))
+			checkContractNegotiationStatus = verifyEDRRequestStatus(offer.getAssetId());
+
+		return checkContractNegotiationStatus;
+	}
+
+	@SneakyThrows
+	public EDRCachedResponse verifyEDRRequestStatus(String assetId) {
+		EDRCachedResponse eDRCachedResponse = null;
+		String edrStatus = "NewToSDE";
+		List<EDRCachedResponse> eDRCachedResponseList = null;
+		int counter = 1;
+		try {
+			do {
+				if (counter > 1)
+					Thread.sleep(THRED_SLEEP_TIME);
+				eDRCachedResponseList = edrRequestHelper.getEDRCachedByAsset(assetId);
+				eDRCachedResponse = verifyEDRResponse(eDRCachedResponseList);
+
+				if (eDRCachedResponse != null)
+					edrStatus = eDRCachedResponse.getEdrState();
+
+				log.info("Verifying 'NEGOTIATED' EDC EDR status to download data for '" + assetId
+						+ "', The current status is '" + edrStatus + "', Attempt " + counter);
+				counter++;
+			} while (counter <= RETRY && !NEGOTIATED.equals(edrStatus));
+
+			if (eDRCachedResponse == null)
+				throw new ServiceException("Time out!! unable to get EDR negotiated status");
+
+		} catch (FeignException e) {
+			log.error("RequestBody: " + e.request());
+			String errorMsg = "FeignExceptionton for asset " + assetId + "," + e.contentUTF8();
+			log.error("Response: " + errorMsg);
+			throw new ServiceException(errorMsg);
+		} catch (InterruptedException ie) {
+			Thread.currentThread().interrupt();
+			String errorMsg = "InterruptedException for asset " + assetId + "," + ie.getMessage();
+			log.error(errorMsg);
+			throw new ServiceException(errorMsg);
+		} catch (Exception e) {
+			String errorMsg = "Exception for asset " + assetId + "," + e.getMessage();
+			log.error(errorMsg);
+			throw new ServiceException(errorMsg);
```

Comment:
consider to make ServiceException be a subclass of RuntimeException. There are many cases where methods which throw  ServiceException are annotated with @SneakyThrows